syntax = "proto3";

package model.protos;

import "model/protos/options.proto";

/* cmd_id == 1 */
message InverterHeartbeat {
    option (homie_node) = { id: "pv1", name: "PV1", type: "pv1"};
    option (homie_node) = { id: "pv2", name: "PV2", type: "pv2" };
    option (homie_node) = { id: "battery", name: "Battery", type: "battery"};
    option (homie_node) = { id: "inverter", name: "Inverter", type: "inverter"};
    option (homie_node) = { id: "llc", name: "LLC", type: "llc"};
    option (homie_node) = { id: "states", name: "States", type: "states"};
    // additional "energy" field with data from EnergyTotalReport is added manually in the code

    option (derived_field) = { 
        operator: 0
        field_name: "pvTotal"
        display_name: "PV total"
        node: "states"
        fields: ["pv1InputWatts", "pv2InputWatts"]
        unit: "W"
    };

    optional uint32 invErrCode = 1 [(mapping_options) = { node: "inverter" }];
    optional uint32 invWarnCode = 3 [(mapping_options) = { node: "inverter" }];
    optional uint32 pv1ErrCode = 2 [(mapping_options) = { node: "pv1" }];
    optional uint32 pv1WarnCode = 4 [(mapping_options) = { node: "pv1" }];
    optional uint32 pv2ErrCode = 5 [(mapping_options) = { node: "pv2" }];
    optional uint32 pv2WarningCode = 6 [(mapping_options) = { node: "pv2" }];
    optional uint32 batErrCode = 7 [(mapping_options) = { node: "battery" }];
    optional uint32 batWarningCode = 8 [(mapping_options) = { node: "battery" }];
    optional uint32 llcErrCode = 9 [(mapping_options) = { node: "llc" }];
    optional uint32 llcWarningCode = 10 [(mapping_options) = { node: "llc" }];
    optional uint32 pv1Status = 11 [(mapping_options) = { node: "pv1" }];
    optional uint32 pv2Status = 12 [(mapping_options) = { node: "pv2" }];
    optional uint32 batStatus = 13 [(mapping_options) = { node: "battery" }];
    optional uint32 llcStatus = 14 [(mapping_options) = { node: "llc" }];
    optional uint32 invStatus = 15 [(mapping_options) = { node: "inverter" }];
    optional int32 pv1InputVolt = 16 [(mapping_options) = { divisor: 10, unit: "V", node: "pv1" }];
    optional int32 pv1OpVolt = 17 [(mapping_options) = { divisor: 100, unit: "V", node: "pv1" }];
    optional int32 pv1InputCur = 18 [(mapping_options) = {divisor: 10, unit: "A", node: "pv1" }];
    optional int32 pv1InputWatts = 19 [(mapping_options) = { divisor: 10, unit: "W", node: "pv1" }];
    optional int32 pv1Temp = 20 [(mapping_options) = {divisor: 10, unit: "°C", node: "pv1" }];
    optional int32 pv2InputVolt = 21[(mapping_options) = {divisor: 10, unit: "V", node: "pv2"}];
    optional int32 pv2OpVolt = 22 [(mapping_options) = {divisor: 100, unit: "V", node: "pv2" }];
    optional int32 pv2InputCur = 23 [(mapping_options) = {divisor: 10, unit: "A", node: "pv2" }];
    optional int32 pv2InputWatts = 24 [(mapping_options) = { divisor: 10, unit: "W", node: "pv2" }];
    optional int32 pv2Temp = 25 [(mapping_options) = {divisor: 10, unit: "°C", node: "pv2" }];
    optional int32 batInputVolt = 26 [(mapping_options) = {divisor: 10, unit: "V", node: "battery" }];
    optional int32 batOpVolt = 27 [(mapping_options) = {divisor: 10, unit: "V", node: "battery" }];
    optional int32 batInputCur = 28 [(mapping_options) = {divisor: 1000, unit: "A", node: "battery" }];
    optional int32 batInputWatts = 29 [(mapping_options) = { divisor: 10, unit: "W", node: "battery" }];
    optional int32 batTemp = 30 [(mapping_options) = {divisor: 10, unit: "°C", node: "battery" }];
    optional uint32 batSoc = 31 [(mapping_options) = {unit: "%", node: "battery" }];
    optional int32 llcInputVolt = 32 [(mapping_options) = {divisor: 10, unit: "V", node: "llc" }];
    optional int32 llcOpVolt = 33 [(mapping_options) = {divisor: 100, unit: "V", node: "llc" }];
    optional int32 llcTemp = 34 [(mapping_options) = {divisor: 10, unit: "°C", node: "llc" }];
    optional int32 invInputVolt = 35 [(mapping_options) = {divisor: 100, unit: "V", node: "inverter" }];
    optional int32 invOpVolt = 36 [(mapping_options) = {divisor: 10, unit: "V", node: "inverter" }];
    optional int32 invOutputCur = 37 [(mapping_options) = {divisor: 1000, unit: "A", node: "inverter" }];
    optional int32 invOutputWatts = 38 [(mapping_options) = { divisor: 10, unit: "W", node: "inverter" }];
    optional int32 invTemp = 39 [(mapping_options) = {divisor: 10, unit: "°C", node: "inverter" }];
    optional int32 invFreq = 40 [(mapping_options) = {divisor: 10, unit: "Hz", node: "inverter" }];
    optional int32 invDcCur = 41 [(mapping_options) = {divisor: 1000, unit: "A", node: "inverter" }];
    optional int32 bpType = 42;
    optional int32 invRelayStatus = 43 [(mapping_options) = { node: "inverter" }];
    optional int32 pv1RelayStatus = 44 [(mapping_options) = { node: "pv1" }];
    optional int32 pv2RelayStatus =  45 [(mapping_options) = { node: "pv2" }];
    optional uint32 installCountry = 46;
    optional uint32 installTown = 47;
    optional uint32 permanentWatts = 48 [(mapping_options) = { divisor: 10, unit: "W", node: "states", settable: true }];
    optional uint32 dynamicWatts = 49 [(mapping_options) = { divisor: 10, unit: "W", node: "states"}];
    optional uint32 supplyPriority = 50 [(mapping_options) = {node: "states", settable: true}];
    optional uint32 lowerLimit = 51 [(mapping_options) = { unit: "%", node: "states", settable: true }];
    optional uint32 upperLimit = 52 [(mapping_options) = { unit: "%", node: "states", settable: true }];
    optional uint32 invOnOff = 53 [(mapping_options) = { node: "inverter" }];
    optional uint32 wirelessErrCode = 54;
    optional uint32 wirelessWarnCode = 55;
    optional uint32 invBrightness = 56 [(mapping_options) = { divisor: 10, unit: "%", node: "inverter", settable: true }];
    optional uint32 heartbeatFrequency = 57;
    optional uint32 ratedPower = 58 [(mapping_options) = { divisor: 10, unit: "W", node: "states" }];
    optional int32 batChargingTime = 59 [(mapping_options) = { converter: "minutes", node: "battery" }];
    optional int32 batDischargingTime = 60 [(mapping_options) = { converter: "minutes", node: "battery" }];
    optional uint32 feedPriority = 61 [(mapping_options) = {node: "states", settable: true}];
}

message InverterHeartbeat2 {
    optional int32 X_Unknown_1 = 1;
    optional int32 X_Unknown_2 = 2;
    optional int32 X_Unknown_3 = 3;
    optional int32 X_Unknown_4 = 4;
    optional int32 X_Unknown_5 = 5;
    optional int32 X_Unknown_6 = 6;
    optional int32 X_Unknown_7 = 7;
    optional int32 X_Unknown_8 = 8;
    optional int32 X_Unknown_9 = 9;
    optional int32 X_Unknown_10 = 10;
    optional int32 X_Unknown_11 = 11;
    optional int32 X_Unknown_12 = 12;
    optional int32 X_Unknown_13 = 13;
    optional int32 X_Unknown_14 = 14;
    optional int32 X_Unknown_15 = 15;
    optional int32 X_Unknown_16 = 16;
    optional int32 X_Unknown_17 = 17;
    optional int32 X_Unknown_18 = 18;
    optional int32 X_Unknown_19 = 19;
    optional int32 X_Unknown_20 = 20;
    optional int32 X_Unknown_21 = 21;
    optional int32 X_Unknown_22 = 22;
    optional int32 X_Unknown_23 = 23;
    optional int32 X_Unknown_24 = 24;
    optional int32 X_Unknown_25 = 25;
    optional int32 X_Unknown_26 = 26;
    optional int32 X_Unknown_27 = 27;
    optional int32 X_Unknown_28 = 28;
    optional int32 X_Unknown_29 = 29;
    optional int32 X_Unknown_30 = 30;
    optional int32 X_Unknown_31 = 31;
    optional int32 X_Unknown_32 = 32;
    optional int32 X_Unknown_33 = 33;
    optional int32 X_Unknown_34 = 34;
    optional int32 X_Unknown_35 = 35;
    optional int32 X_Unknown_36 = 36;
    optional int32 X_Unknown_37 = 37;
    optional int32 X_Unknown_38 = 38;
    optional int32 X_Unknown_39 = 39;
    optional int32 X_Unknown_40 = 40;
    optional int32 X_Unknown_41 = 41;
    optional int32 X_Unknown_42 = 42;
    optional int32 X_Unknown_43 = 43;
    optional int32 X_Unknown_44 = 44;
    optional int32 X_Unknown_45 = 45;
    optional int32 X_Unknown_46 = 46;
    optional int32 X_Unknown_47 = 47;
    optional int32 X_Unknown_48 = 48;
    optional int32 X_Unknown_49 = 49;
    optional int32 X_Unknown_50 = 50;
    optional int32 X_Unknown_51 = 51;
    optional int32 X_Unknown_52 = 52;
 }

message Header {
	optional bytes pdata = 1;
    optional int32 src = 2;
    optional int32 dest = 3;
    optional int32 d_src= 4;
    optional int32 d_dest = 5;
    optional int32 enc_type = 6;
    optional int32 check_type = 7;
    optional int32 cmd_func = 8;
    optional int32 cmd_id = 9;
    optional int32 data_len = 10;
    optional int32 need_ack = 11;
    optional int32 is_ack = 12;
    optional int32 seq = 14;
    optional int32 product_id = 15;
    optional int32 version = 16;
    optional int32 payload_ver = 17;
    optional int32 time_snap = 18;
    optional int32 is_rw_cmd = 19;
    optional int32 is_queue = 20;
    optional int32 ack_type= 21;
    optional string code = 22;
    optional string from = 23;
    optional string module_sn = 24;
    optional string device_sn = 25;

}

message SendHeaderMsg {
    repeated Header msg = 1;
}

message SetMessage {
    Header msg = 1;
}

message SetValue {
    optional int32 value = 1;
}